doc/decisions/DR023-Versioning-traits-and-specifications.md

```diff
@@ -0,0 +1,702 @@
+  version. (Although, hosts _could_ query managers for what version of a
+  trait they should send via a version aware `managementPolicy` or some
+  other mechanism, so this is perhaps a weak argument. That arrangement
+  seem intuitively more complex though)
+- In many cases, manager plugins are privately owned by pipeline
+  authors, who are also in control of when to update the hosts in their
+  pipeline.
+
+## Manager/Host Namespace Collisions
+With the current state of affairs, a large problem with versioning is
+presented when you consider the possibility of namespace collision
+between hosts and managers.
+
+For the purpose of this decision, we will assume that two versions of
+`mediacreation` can exist in the same process without problematic
+namespace collisions, although this remains a problem to be solved.
+
+(Note: Practically, this may mean that we end up strongly recommending
+ vendoring mediacreation into manager plugins, especially for python
+ libraries which can't do C++ tricks with private symbols.)
+
+## Schema version, package version
+
+The schema version (the version number in the `.yml`) and the package
+version (the `openassetio-mediacreation` library version) are separate.
+
+It is possible that updates to `openassetio` or `openassetio-traitgen`
+could cause breakages that necessitate a major version update to the
+`openassetio-mediacreation` package.
+
+We are assuming that this does **not** imply a schema version update.
+
+
+# What to Version
+In considering how to version the openassetio traits and specifications,
+two options immediately present themselves. Version every trait and
+specification individually, or version the entire schema as a whole.
+
+## Version every trait
+Every trait gets its own disconnected version. This version would have
+to travel across the api-boundary along with the trait data. Whilst the obvious
+thing would be to enhance the low level `TraitsData` type to have a
+first class understanding of versioning, this cannot be so because not
+every method takes a `TraitsData` as an argument. The lowest-friction
+course of action might be to pack the version into `TraitId`, essentially
+doing string-munging to communicate the version across the api-boundary.
+
+Hosts and managers would need to be inspect the version data, which we
+could provide conveniences for, and construct the correct version
+of the trait view.
+
+```yml
+# traits.yml
+Camera:
+  description: >
+    The entity represents a device that captures a view through
+    re-projection or encoding into some other form.
+  usage:
+    - entity
+  version: 3
+```
+
+```python
+# Manager
+# In this example, CameraTrait is on version 3
+# SpatialTrait is on version 2
+#
+# The manager is responding to a resolve, needing to populate a
+# TraitsData with a camera trait and a spatial trait of the expected
+# version
+cameraVersion = mediacreation.versionFromTraitId(mediacreation.CameraTrait.kId)
```

Comment:
[question] Would the version be determined from the constant from the runtime package, or the requested trait ID in the `resolve` request.  The latter may be beneficial in case this is running out-of-process (eg: via gRPC) where the imported `openassetio-mediacreation` package may not be the version the host has imported.

Reply:
In this example it was coming from the traitID itself yes, (essentially we were thinking of appending/prepending it to the string.)

I don't think there's any future in determining the version from the runtime package itself, for exactly the reasons you say. The host will have to inform the manager somehow of the version directly.

Reply:
> ...essentially we were thinking of appending/prepending it to the string....

> ...The host will have to inform the manager somehow of the version directly, ...

If its in the id string then do you need another mechanism?

Reply:
That's a really good argument to encoding version in the trait ID - also good arguments in https://github.com/OpenAssetIO/OpenAssetIO/pull/1254#discussion_r1482555651